File: leetcode/295.数据流的中位数/295.数据流的中位数.go
package main

import (
	"container/heap"
	"sort"
)

/*
定义类型，用Sort.InSlice做容器，实现heap的Push和Pop接口实现heap功能，完成大小堆的使用
heap 是最小堆，所以最大堆是用负值来实现
注意：最大堆的值和到最小堆中的符号转换
*/
type hp struct {
	sort.IntSlice
}

func (h *hp) Push(v interface{}) {
	h.IntSlice = append(h.IntSlice, v.(int))
}
func (h *hp) Pop() interface{} {
	a := h.IntSlice
	v := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return v
}

type MedianFinder struct {
	// small是最大堆，放是小值
	// large是最小堆，放的大值
	// large和small对接的时候 就是中位数的位置
	small, large *hp
}

/** initialize your data structure here. */
func Constructor() MedianFinder {
	return MedianFinder{
		small: &hp{},
		large: &hp{},
	}
}

func (this *MedianFinder) AddNum(num int) {
	if this.small.Len() >= this.large.Len() {
		heap.Push(this.small, -num)
		heap.Push(this.large, -heap.Pop(this.small).(int))
	} else {
		heap.Push(this.large, num)
		heap.Push(this.small, -heap.Pop(this.large).(int))
	}

}

func (this *MedianFinder) FindMedian() float64 {
	if this.small.Len() > this.large.Len() {
		return float64(-this.small.IntSlice[0])
	} else if this.large.Len() > this.small.Len() {
		return float64(this.large.IntSlice[0])
	}
	// float/2 结果是float，int/2 还是int
	return float64(-this.small.IntSlice[0]+this.large.IntSlice[0]) / 2
}

func main() {
	mf := Constructor()
	mf.AddNum(1)
	mf.AddNum(2)
	println(mf.FindMedian())
	mf.AddNum(3)
	println(mf.FindMedian())
}

File: leetcode/295.数据流的中位数/295.数据流的中位数.py
import heapq


class MedianFinder:

    def __init__(self):
        """
        large列表：小顶堆
        small列表：大顶堆
        heapq.heappush()是往堆中添加新值，此时自动建立了小顶堆
        创建大顶堆每次push时给元素加一个负号
        """
        self.large = list()
        self.small = list()

    def addNum(self, num: int) -> None:
        """
        添加元素，使用堆的方法添加
        谁元素少，就加到它那里（隔山打牛)
            - 先把num添加到 元素多的堆里
            - 再把元素多的堆 对顶添加到元素少的堆里面
        注意：大顶堆值 负号的问题
        """
        small_ = self.small
        large_ = self.large

        if len(small_) >= len(large_):
            heapq.heappush(small_, -num)
            heapq.heappush(large_, -heapq.heappop(small_))
        else:
            heapq.heappush(large_, num)
            heapq.heappush(small_, -heapq.heappop(large_))

    def findMedian(self) -> float:
        """
        查找中值，不需要pop
        谁元素多，谁就是中值
        一样多，顶值相加 / 2
        """
        if len(self.small) > len(self.large):
            return -self.small[0]
        elif len(self.small) < len(self.large):
            return self.large[0]
        return (-self.small[0] + self.large[0]) / 2


if __name__ == '__main__':
    obj = MedianFinder()
    obj.addNum(1)
    obj.addNum(2)
    param_2 = obj.findMedian()
    print(param_2)
    obj.addNum(3)
    param_3 = obj.findMedian()
    print(param_3)
